### PR TITLE
Cat-log - open job logs in editor

### PR DESCRIPTION
**Pull request by challurip**
`cylc cat-log` - allow opening job logs in users' editor ( temporary read only copy ); same via gcylc.

~~[not ready for review yet - coming soon]~~

### CODE REVIEW

**Comment by oliver-sanders** [comment]
This will be a really useful feature! Just taking a quick glance through before review:

Something which would make it more convenient is if the editor command is run in the background in the case that the `--geditor` option is set. That way users could continue to work in the terminal whilst going through the log file.
_`Popen` runs commands in the background by default its the `proc.communicate` and `proc.wait` methods which cause the process to wait until the command returns._

PS: There is a typo in `bin/cylc-cat-log` at line 302.

File: bin/cylc-cat-log [hunk 1]
```diff
@@ -271,6 +284,32 @@ def main():
...
+    if options.editor or options.geditor:
+        if os.path.exists(filename):
+            lines = read_and_proc(filename)
+        else:
+            print >> sys.stderr, 'No such file or directory: %s' % filename
+            sys.exit(1)
+
+        viewfile = NamedTemporaryFile(
+            suffix='.' + os.path.basename(filename),
+            prefix=suite + '.', dir=cylc_tmpdiir
```
Comment (hjoliver) [inline bin/cylc-cat-log]: typo (`diir`) - spotted by @oliver-sanders

File: bin/cylc-cat-log [hunk 2]
```diff
@@ -271,6 +284,32 @@ def main():
...
+        )
+
+        for line in lines:
+            viewfile.write(line + '\n')
+            viewfile.seek(0, 0)
```
Comment (hjoliver) [inline bin/cylc-cat-log]: Indentation error - `seek` after all lines written, not after each.

File: bin/cylc-cat-log [hunk 3]
```diff
@@ -309,13 +352,31 @@ def main():
...
+                sys.exit(1)
+
+            modtime2 = os.stat(viewfile.name).st_mtime
+
+            if modtime2 > modtime1:
+                print >> sys.stderr, (
+                    'WARNING: YOU HAVE EDITED A TEMPORARY READ_ONLY COPY OF: ')
+                print >> sys.stderr, filename
+            else:
+                print >> sys.stderr, 'NO CHANGES MADE TO: %s' % viewfile.name
```
Comment (hjoliver) [inline bin/cylc-cat-log]: We don't need this line IMO.

**Comment by hjoliver** [comment]
@oliver-sanders - 

> Something which would make it more convenient is if the editor command is run in the background in the case that the --geditor option is set. 

The reason we block on the editor command is to check and warn if the user edited and force-saved the file, which suggests they thought they were editing the original file rather than a temporary read-only copy. (This is more important in `cylc view --geditor` which views the suite.rc file, but I think it's probably worth doing here too).  I think the right way to do what you want is:

```
cylc cat-log --geditor my-suite my-task.2010 &
```

(most GUI editors - except for `gvim` default behaviour - don't automatically background either btw - i.e. you need to use `&` on the command line).

**Comment by challurip** [comment]
Made changes as suggested.

File: bin/cylc-cat-log [hunk 4]
```diff
@@ -271,6 +284,32 @@ def main():
...
+    if options.editor or options.geditor:
+        if os.path.exists(filename):
+            lines = read_and_proc(filename)
+        else:
+            print >> sys.stderr, 'No such file or directory: %s' % filename
+            sys.exit(1)
+
+        viewfile = NamedTemporaryFile(
+            suffix='.' + os.path.basename(filename),
+            prefix=suite + '.', dir=cylc_tmpdir
```
Comment (hjoliver) [inline bin/cylc-cat-log]: One more problem found while testing: `prefix=suite` doesn't work for suite names containing `/` corresponding to the directory path under `cylc-run/` - this makes the viewfile path refer to a non-existent directory.  Can you change to this:
`prefix=suite.replace('/','_') + '.'` - and also do the same for the original code in `cylc view`.

File: bin/cylc-cat-log [hunk 5]
```diff
@@ -296,6 +335,10 @@ def main():
             cmd_tmpl = str(GLOBAL_CFG.get_host_item(
                 "local tail command template"))
             commands.append(shlex.split(cmd_tmpl % {"filename": filename}))
+    elif options.geditor or options.editor:
+        command_list = editor.split()
```
Comment (hjoliver) [inline bin/cylc-cat-log]: @challurip - if you can make this change and push it up today, I'll merge this for the new release (tonight, if the anesthetic has worn off...)

**Pull request merged**